Documentation for Solidity Smart Contracts (#15)
xlink-network/xlink-docs: developers/contracts/BridgeEndpoint.md

# BridgeEndpoint
- Location: `xlink/packages/contracts/bridge-solidity/contracts`

> - Deployed contracts: [BridgeEndpoint](0x84254dA34abE4678017A5Bf78506B48490ce4547), [BridgeEndpointWithSwap](0x84254dA34abE4678017A5Bf78506B48490ce4547).

Reviewer: Los links no funcionan. En general, cuando se referencia un contrato usando una URL, se hace poniendo el link de un blockchain explorer. Ejemplo: https://etherscan.io/address/0x84254dA34abE4678017A5Bf78506B48490ce4547.

Por otro lado, como el `BridgeEndpointWithSwap` extiende a `BridgeEndpoint`, una vez que se deploya BEWS, reemplaza al BE que había antes. Es decir, quedarán como dos contratos distintos en la blockchain, cada uno con su address, pero pasará a usarse el BEWS. Hacen un upgrade de facto: de ahora en más se usa la nueva address. Es **muy** importante que entiendas este concepto. Si te quedan dudas, porfas consultá en el canal de Slack.

En este caso particular de EVM, como hay muchas blockchains que son EVM-compatible y donde ALEX tiene contratos deployados, diría que referenciemos https://docs.xlink.network/xlink-network/readme/ethereum-contract-addresses y listo. Es decir, poner algo de este estilo:

```
- Deployed contracts: See [Ethereum Contract Addresses](https://docs.xlink.network/xlink-network/readme/ethereum-contract-addresses).
```

Para que quienes estés interesado en las direcciones de los contratos deployados, miren la tabla y listo.

Author: Reemplace los links por el de xlink network.

> This technical document provides a detailed overview of the Bridge Endpoint in the Ethereum blockchain. The Bridge Endpoint facilitates communication between two blockchain networks by acting as the entry and exit point for assets moving along the Cross Chain Bridge. It passes messages between chains in the form of events, triggers contract calls, processes token transfers and validates and executes the unwrapping of tokens. The Bridge Endpoint functionality is implemented across two contracts, `BridgeEndpoint` and `BridgeEndpointWithSwap`. `BridgeEndpointWithSwap` extends `BridgeEndpoint` and implements the necessary functionality to source liquidity from external aggregators.

Reviewer: > This technical document provides a detailed overview of the Bridge Endpoint **on EVM-compatible blockchains.**

Notar que no es sólo para Ethereum, sino para muchas blockchains que usan la Ethereum Virtual Machine (todas las que están listadas [acá](https://docs.xlink.network/xlink-network/readme/ethereum-contract-addresses#table-of-contents)).

This functionality is implemented and distributed across the following contracts:

- `BridgeEndpoint`: the base contract that facilitates bridging operations.
- `BridgeEndpointWithSwap`: extends `BridgeEndpoint` and integrates swaps during a bridge transfer.

## Storage

### `registry`

| Data     | Type   |
| -------- | ------ |
| Variable | `BridgeRegistry` |

Stores a reference to the `BridgeRegistry` contract, which manages approved tokens, relayers, and validators.

### `pegInAddress`

| Data     | Type   |
| -------- | ------ |
| Variable | `address` |

> The address where tokens are deposited before bridging.

Reviewer: Desde mi POV, "before bridging" es medio indefinido. Qué significa? En qué casos se usa la `pegInAddress`?

Propuesta (sin conocer a fondo el contrato, puede no ser del todo correcta):
> The address where tokens are deposited **during peg-out operations and from which they are transferred during peg-in operations**.

### `timeLock`

| Data     | Type   |
| -------- | ------ |
| Variable | `ITimeLock` |

> Manages locked transactions that require a delay before execution. Tokens exceeding the timelock threshold are not immediately sent to the user. Instead, the timeLock contract holds them until the waiting period expires and then releases the tokens.

Reviewer: 1. Me resulta confuso "Tokens exceeding the timelock threshold [...]", no sería "**Peg-in amounts** exceeding the timelock threshold"?

2. Qué sería "releases the tokens"? Se los envía al usiario? El usuario puede pedirlos y se los dan?

3. Backticks para timeLock en la tercera oración.

Author: Aclare lo de los tokens, releasing the tokens seria completar la transaccion pendiente, que no se ejecuto. Es decir, se los mandaria al usuario como en cualquier transaccion, solo que despues de un tiempo, segun entiendo.

> ### `timeLockThreshold`

Reviewer: Este es un timelock treshold global? Viendo https://github.com/xlink-network/xlink/blob/main/packages/contracts/bridge-solidity/contracts/BridgeEndpoint.sol#L224, entiendo que sí. 

Sugerencia: estaría bueno que esto se entienda, a lo mejor se puede aclarar en `timeLockThresholdByToken`, diciendo que la lógica del timelock se va a ejecutar cuando el amount supere el `timeLockThreshold` y el custom, si existiese.

Author: Es global, actualice la explicacion

| Data     | Type   |
| -------- | ------ |
| Variable | `uint256` |

The minimum token amount that triggers a timelock. By default, the value is set to `0`.

### `timeLockThresholdByToken`
| Data     | Type   |
| -------- | ------ |
| Variable | `mapping(address => uint256)` |

Custom timelock thresholds for different tokens.

### `unwrapSent`
| Data     | Type   |
| -------- | ------ |
| Variable | `mapping(bytes32 => OrderPackage)` |

Stores unwrap transactions that need to be finalized. It stores a mapping of `OrderPackage` structs, which contain a flag indicating whether the unwrap has been completed.

### `swapExecutor`
###### _(only present in BridgeEndpointWithSwap)_

| Data     | Type   |
| -------- | ------ |
| Variable | `SwapExecutor` |

> Holds a reference to SwapExecutor, which is the contract that executes swaps.

Reviewer: Agregar backticks en SwapExecutor.

### `swapSent`
###### _(only present in BridgeEndpointWithSwap)_

| Data     | Type   |
| -------- | ------ |
| Variable | `mapping(bytes32 => SwapOrderPackage)` |

> A mapping to track swap orders. `bytes32` is a unique hash of the swap parameters, as calculated by `transferToSwap`.

Reviewer: Me copa esto de explicar que la key `bytes32` del mapping es el hash de los parámetros del swap. Buscaría consistencia entre esta descripción y la de `unwrapSent` dado que son muy análogas. Por ejemplo: agregar eso del hash.

Por otro lado, si se menciona alguna función, está bueno poner un markdown link ``[`transferToSwap`](#transfer-to-swap)``.

Author: Done. Aplique links siempre que se menciona algo que esta definido en el documento

## Data Types

#### `OrderPackage`

Holds details of a pending unwrap operation.

```solidity
struct OrderPackage {
  address recipient;
  address token;
  uint256 amount;
  bool sent;
}
```

#### `SignaturePackage`

> Contains signatures from validators to prove a transaction is legitimate. It is used for verifying cross-chain transactions.

Reviewer: > Contains signatures from validators to prove **an order** is legitimate.

```solidity
struct SignaturePackage {
  bytes32 orderHash;
  address signer;
  bytes signature;
}
```

#### `SwapOrderPackage`
###### _(only present in BridgeEndpointWithSwap)_

A struct that stores details of a swap transaction.

```solidity
struct SwapOrderPackage {
  address target;
  address tokenIn;
  address tokenOut;
  uint256 amountIn;
  uint256 amountOutMin;
  bytes bridgePayloadSuccess;
  bytes bridgePayloadFailure;
  bool sent;
}
```

## Modifiers

- `onlyApprovedToken(token)`: ensures the token is approved in the registry.
- `onlyApprovedRelayer()`: ensures the caller is an approved relayer.
- `notWatchlist(recipient)`: prevents transactions to watchlisted addresses.
- `nonReentrant`: protects against reentrancy attacks.
- `onlyAllowlisted`: ensures only allowed addresses can execute certain functions.

## Features

#### `sendMessageWithToken`

> This function transfers a token across chains, deducts a fee, and emits an event. The user deposits tokens on the bridge contract in the source chain. The contract checks that the token is approved (since `sendMessageWithToken` has the `onlyApprovedToken` modifier). This function calls `_transfer`, which performs validations. Finally, it emits a `SendMessageWithTokenEvent` containing the transaction details. Afterwards, off-chain validators sign the transaction.

Reviewer: 1. Cuál es la source chain? Siento que esta función está descrita como "mirándola de lejos".  Esta función es la que ejecuta los peg-outs, sirve para enviar tokens desde la EVM chain a otras blockchains. 

2. En lugar de poner "This function calls `_transfer`, which performs validations.", me parece más útil, o bien mencionar algunas de las validaciones que hace, o bien poner a `_transfer` dentro de una sección de relevant internal functions (ver [acá](https://github.com/CoinFabrik/alex-qa-internal/wiki/Templates-Contracts#v011)). 

3. Me parece importante no confundir _transaction_ con _order_. En general, cuando hablamos de transacciones lo hacemos en el marco de una blockchain (las transacciones son las que cambian el _world state_, el estado del sistema en el que estamos trabajando, la blockchain). Cuando hablamos de órdenes, en el marco de ALEX y XLink, en general estamos hablando de órdenes del bridge. Si alguien quiere pasar tokens de Arbitrum a Stacks, decimos que se genera una orden. Para Arbitrum, va a ser una orden de peg-out. Para Stacks, es una order de peg-in. Para que la plata llegue de un lado a otro se necesita una infraestructura off-chain que es la que provee XLink con workers (relayers y validadores).

Entonces, el usuario ejecuta una transacción llamando a la función `sendMessageWithToken`. En esa transacción, el usuario transfiere los tokens que quiere bridgear al contrato endpoint. En esa transacción, también se genera un evento. Este evento es escuchado por los relayers y los validadores para armar y validar la orden iniciada por el usuario. El validador debe corroborar que la transacción efectivamente exisitió y que el usuario depositó los tokens en el bridge endpoint. Una vez que lo verifica, firma la orden.

Todo esto lo escribo no para que vaya en la descripción, sino para que quede más claro el concepto y puedas modificar acorde.

Reviewer: Con respecto a 2.:

Veo que `_transfer` está como relevant internal functions! Genial. Entonces hay que referenciarla.

Author: Esta "mejorada" la explicacion, no entre en detalle sobre transfer porque esta explicada en internal functions, solo la referencie. Cambie transaction por order (o viceversa) donde estaba mal usado alguno de los dos conceptos.

Reviewer: Muy buena!!

##### Parameters
```solidity
address token, uint256 amount, bytes calldata payload
```

#### `sendMessage`

Emits an event with a message.

##### Parameters
```solidity
bytes calldata payload
```

#### `transferToUnwrap`

> This function is called by a relayer after `sendMessageWithToken` is executed, with the purpose of sending tokens to a recipient. Relayers will listen for the corresponding `SendMessageWithTokenEvent` on the source chain and call `transferToUnwrap` providing the token, recipient and amount, a salt (a unique transaction identifier, usually the source chain transaction hash), and an array of validator signatures (proofs). The contract validates the transaction and generates a EIP-712 hash, which acts as a unique identifier for each swap.

Reviewer: > This function is called by a relayer after `sendMessageWithToken` is executed.

No necesariamente. Si sucede, `sendMessageWithToken` fue llamada desde otra EVM-chain. La función es llamada por un relayer cuando el usuario inició una orden en otra blockchain. Esa otra blockchain puede ser Stacks, Solana u otra EVM. 

> The contract validates the transaction and generates a EIP-712 hash, which acts as a unique identifier for each swap.

For each hash?

Author: Cada swap tiene su propio hash. La primera parte la corregi pero es la correccion que menos segura me deja.

##### Parameters
```solidity
address token,
address recipient,
uint256 amount,
bytes32 salt,
SignaturePackage[] calldata proofs
```

#### `finalizeUnwrap`

> Completes an unwrap transaction and releases tokens to the recipient. If the token transfer was delayed (timelocked), the unwrap must be manually finalized. The function loops through each `orderHash` and calls `_finalizeUnwrap(orderHash)`, which verifies that the transaction has not already been completed, transfers tokens from `unwrapSent` mapping to the recipient and marks the order as sent. Finally, it emits `FinalizeUnwrapEvent.`

Reviewer: > [...] transfers tokens from `unwrapSent` mapping to the recipient [...] 

Desde dónde se mandan los tokens? Quién llama a `finalizeUnwrap`?

##### Parameters
```solidity
address target,
address tokenIn,
address tokenOut,
uint256 amountIn,
uint256 amountOutMin,
bytes calldata swapPayload,
bytes calldata bridgePayloadSuccess,
bytes calldata bridgePayloadFailure,
bytes32 salt,
SignaturePackage[] calldata proofs
```

#### `finalizeSwap`
###### _(in contract BridgeEndpointWithSwap)_`
This function is used when a token is not burnable, leading `transferToSwap` to store the swap order instead of executing it immediately. It ensures input arrays are valid, and it loops through each `orderHash` and calls `_finalizeSwap()`.

##### Parameters
```solidity
bytes32[] calldata orderHashes,
bytes[] calldata swapPayloads
```

### Governance Features

#### `setTimeLock`

Updates the contract managing timelocks.

##### Parameters
```solidity
address _timeLock
```
#### `setTimeLockThreshold`

Sets the global timelock threshold.

##### Parameters
```solidity
uint256 _timeLockThreshold
```

#### `setTimeLockThresholdByToken`

Sets a custom timelock threshold per token.

##### Parameters
```solidity
address token, uint256 _timeLockThreshold
```

> #### `addAllowList`

Reviewer: Faltan parámetros acá.

Adds an address to the allowlist, granting it permission to perform specific contract actions.

#### `removeAllowedList`

Removes an address from the allowlist, revoking its access.

##### Parameters
```solidity
address account
```

#### `pause`
Pauses the contract, preventing token transfers until the contract is unpaused.

#### `unpause`
Resumes contract operations after a pause, allowing bridging and transfers again.

### Read-Only Functions

> #### `onAllowList`

Reviewer: Faltan parámetros.

Returns `true` if the provided address is on the allowlist, which means it has permission to use the contract’s functions.

#### `offAllowList`

Returns `true` if the provided address is not on the allowlist. 

##### Parameters
```solidity
address account
```

### Relevant Internal Functions

#### `_transfer`

> Transfers a token from the sender, deducts fees, and sends the remainder to the bridge.

Reviewer: Estaría bueno dar un poco más de info sobre esta función.

- Hace validaciones? Cuáles?
- Quién es el sender? `msg.sender`?
- Qué pasa en el caso de que el token sea burnable? También se transfiere?
- El fee a qué address va? Y el amount a bridgear? De qué address a qué address son las transferencias?

##### Parameters
```solidity
bytes32 orderHash, bytes memory swapPayload
```

#### `_executeSwap`
###### _(in contract BridgeEndpointWithSwap)_`

This function approves `swapExecutor` to spend `tokenIn` and calls its `executeSwap()` function to attempt the swap. If the swap succeeds, it either burns the swapped tokens or prepares them for transfer. To transfer the tokens, `tokenOut` is sent to `pegInAddress` for bridging and a `SendMessageWithTokenEvent` is emitted. If the swap fails, the error is logged via `SwapExecutorError`, approvals are revoked and a `SendMessageWithTokenEvent` with `bridgePayloadFailure` is emitted. In either case, the function will burn `tokenIn` tokens if applicable. 

##### Parameters
```solidity
address tokenIn,
address tokenOut,
address target,
uint256 amountIn,
uint256 amountOutMin,
bytes memory swapPayload,
bytes memory bridgePayloadSuccess,
bytes memory bridgePayloadFailure
```

## Events

#### `SendMessageEvent`

Emitted when a user sends a message without transferring tokens. 

##### Parameters
```solidity
address indexed from, uint256 value, bytes payload 
```

#### `SendMessageWithTokenEvent`

> Emitted when a user executes a transaction.

Reviewer: ```suggestion
Emitted when a user initiates a peg-out order.
```

Emitted when a new swap order is created for non-burnable tokens. It logs the creation of swap orders, helping track orders before execution.

##### Parameters
```solidity
bytes32 indexed orderHash,
address indexed target,
address indexed tokenIn,
address tokenOut,
uint256 amountIn,
uint256 amountOutMin,
bytes bridgePayloadSuccess,
bytes bridgePayloadFailure
```

#### `SwapOrderFinalized`
###### _(only present in BridgeEndpointWithSwap)_

Emitted when a swap order is executed and finalized, confirming a swap has been processed.

##### Parameters
```solidity
bytes32 indexed orderHash,
address indexed executor,
uint256 amountOut,
bool success
```

#### `TransferToSwapEvent`
###### _(only present in BridgeEndpointWithSwap)_

Emitted when a token transfer and swap operation is executed.

##### Parameters
```solidity
bytes32 orderHash,
address target,
bytes swapPayload,
address tokenIn,
address tokenOut,
uint256 amountIn,
uint256 amountOut,
bool success
```

## Contract Calls (Interactions)

- `BridgeRegistry`: this contract is called to process orders and to manage validator roles and fees. It acts as the central registry for approved tokens, relayers and validators. 

> - `ITimeLock`: provides the implementation for timelocked transactions. 

Reviewer: Está bueno explicar cuándo se interactúa con el contrato que tiene esa interfaz.

```suggestion
- `ITimeLock`: the `ITimeLock` interface is utilized to interact with the `timeLock` contract by calling the `createAgreement` function when birdged amounts exceeds the threshold.
```

> - `IBurnable`: handles the wrapping/unwrapping process in a bridge by burning tokens on one chain and minting them on another.

Reviewer: Lo mismo acá. Además, falta la interacción con los tokens `ERC20`. Dejo sugerencia:

```suggestion
- `IBurnable`: this interface is used for burnable tokens to enable mint and burn operations.
- `ERC20`:  all approved tokens within the bridge must implement the `ERC20Fixed` standard, which is an XLink's custom standard that extends ERC-20 to handle fixed precision. Token contract interactions occur in both peg-in and peg-out operations for non-burnable tokens, using the `transferFromFixed`, `transferFixed`, and `increaseAllowanceFixed` functions.
```

# BridgeEndpoint
- Location: `xlink/packages/contracts/bridge-solidity/contracts`
- Deployed contracts: See [Ethereum Contract Addresses](https://docs.xlink.network/xlink-network/readme/ethereum-contract-addresses).

> This technical document provides a detailed overview of the Bridge Endpoint in EVM-compatible blockchains. The Bridge Endpoint facilitates communication between two blockchain networks by acting as the entry and exit point for assets moving along the Cross Chain Bridge. It passes messages between chains in the form of events, triggers contract calls, processes token transfers and validates and executes the unwrapping of tokens. The Bridge Endpoint functionality is implemented across two contracts, `BridgeEndpoint` and `BridgeEndpointWithSwap`. `BridgeEndpointWithSwap` extends `BridgeEndpoint` and implements the necessary functionality to source liquidity from external aggregators.

Reviewer: Revisar esta descripción, queda medio redundante las ultimas oraciones de este párrafo y luego lo de las lineas 7-10. Lo pondría sólo en un lugar.

This functionality is implemented and distributed across the following contracts:

- `BridgeEndpoint`: the base contract that facilitates bridging operations.
- `BridgeEndpointWithSwap`: extends `BridgeEndpoint` and integrates swaps during a bridge transfer.

## Storage

### `registry`

| Data     | Type   |
| -------- | ------ |
| Variable | `BridgeRegistry` |

Stores a reference to the `BridgeRegistry` contract, which manages approved tokens, relayers, and validators.

### `pegInAddress`

| Data     | Type   |
| -------- | ------ |
| Variable | `address` |

> The address from which non-burnable tokens are transferred during peg-in operations.

Reviewer: Esta descripción está mucho mejor, pero sigue incompleta/incorrecta. Te dejo para que la mejores, vale preguntar y mirar los diagramas de la reunión del otro día.

### `timeLock`

| Data     | Type   |
| -------- | ------ |
| Variable | `ITimeLock` |

> Manages locked transactions that require a delay before execution. Tokens amounts that exceed the threshold are not immediately sent to the user. Instead, the `timeLock` contract holds them until the waiting period expires and then releases the tokens, completing the transfer.

Reviewer: ```suggestion
Manages locked transactions that require a delay before execution. Tokens amounts that exceed the threshold are not immediately sent to the user. Instead, the `timeLock` contract holds them until the waiting period expires. At that point, the `timeLock` owner can fulfill the order, completing the cross-chain transfer.
```

## Data Types

#### `OrderPackage`

Holds details of a pending unwrap operation.

```solidity
struct OrderPackage {
  address recipient;
  address token;
  uint256 amount;
  bool sent;
}
```

#### `SignaturePackage`

Contains signatures from validators to prove an order is legitimate. It is used for verifying cross-chain transfers.

```solidity
struct SignaturePackage {
  bytes32 orderHash;
  address signer;
  bytes signature;
}
```

#### `SwapOrderPackage`
###### _(only present in BridgeEndpointWithSwap)_

A struct that stores details of a swap order.

```solidity
struct SwapOrderPackage {
  address target;
  address tokenIn;
  address tokenOut;
  uint256 amountIn;
  uint256 amountOutMin;
  bytes bridgePayloadSuccess;
  bytes bridgePayloadFailure;
  bool sent;
}
```

## Modifiers

- `onlyApprovedToken(token)`: ensures the token is approved in the registry.
- `onlyApprovedRelayer()`: ensures the caller is an approved relayer.
- `notWatchlist(recipient)`: prevents transfers to watchlisted addresses.
- `nonReentrant`: protects against reentrancy attacks.
- `onlyAllowlisted`: ensures only allowed addresses can execute certain functions.

## Features

#### `sendMessageWithToken`

> This function is called to initiate the peg-out process from an EVM-compatible blockchain onto other chains, such as Stacks or Solana. The user deposits tokens in the bridge contract, which are burned or locked, depending on the token. The contract checks that the token is approved, since [`sendMessageWithToken`](#sendmessagewithtoken) has the `onlyApprovedToken` modifier. The function calls [`_transfer`](#_transfer), which performs validations. Finally, the function emits a [`SendMessageWithTokenEvent`](#sendmessagewithtokenevent) containing the transaction details, which will be reviewed by validators. Once they verify that the tokens were deposited in the `BridgeEndpoint` contract, the validators will sign the order and relayers will submit it in the destination chain.

Reviewer: ```suggestion
This function is called to initiate the peg-out process from an EVM-compatible blockchain onto other chains, such as Stacks or Bitcoin. The user deposits tokens in the bridge contract, which are burned or locked, depending on the token. The contract checks that the token is approved, since [`sendMessageWithToken`](#sendmessagewithtoken) has the `onlyApprovedToken` modifier. The function calls [`_transfer`](#_transfer), which performs validations. Finally, the function emits a [`SendMessageWithTokenEvent`](#sendmessagewithtokenevent) containing the transaction details, which will be reviewed by validators. Once they verify that the tokens were deposited in the `BridgeEndpoint` contract, the validators will sign the order and relayers will submit it in the destination chain.
```

##### Parameters
```solidity
address token,
uint256 amount, 
bytes calldata payload
```

#### `sendMessage`

Emits an event with a message.

##### Parameters
```solidity
bytes calldata payload
```

#### `transferToUnwrap`

> This function is called by a relayer when a user initiates a token transfer from another blockchain. The originating order may come from an EVM chain (via [`sendMessageWithToken`](#sendmessagewithtoken)), or from a non-EVM chain like Stacks or Solana. Relayers listen for the corresponding event on the source chain and call [`transferToUnwrap`](#transfertounwrap) on the destination chain, supplying the recipient, token, amount, a salt (usually the source chain order hash), and an array of validator signatures (proofs). The contract verifies these proofs and generates an EIP-712-compliant hash, which acts as a unique identifier for each transfer.

Reviewer: ```suggestion
This function is called by a relayer when a user initiates a token transfer from another blockchain. The originating order may come from an EVM chain (via [`sendMessageWithToken`](#sendmessagewithtoken)), or from a non-EVM chain like Stacks, Bitcoin or Solana. Relayers listen for the corresponding event on the source chain and call [`transferToUnwrap`](#transfertounwrap) on the destination chain, supplying the recipient, token, amount, a salt (usually the source chain transaction), and an array of validator signatures (proofs). The contract verifies these proofs and generates an EIP-712-compliant hash, which acts as a unique identifier for each order.
```

##### Parameters
```solidity
address token,
address recipient,
uint256 amount,
bytes32 salt,
SignaturePackage[] calldata proofs
```

#### `finalizeUnwrap`

> This function completes a pending unwrap order for non-burnable tokens. It is called by a relayer once the timeLock period expires and the recipient can receive their locked tokens. It loops through each `orderHash` and calls [`_finalizeUnwrap()`](#_finalizeunwrap), which verifies that the order has not already been completed and transfers the token and amount stored in [`unwrapSent`](#unwrapsent) to the recipient. The order is then marked as completed, and the [`FinalizeUnwrapEvent`](#finalizeunwrapevent) is emitted.

Reviewer: ```suggestion
This function completes a pending unwrap order for non-burnable tokens. It is called by a hot wallet address once the timeLock period expires, which transfers the tokens to the recipients and finalizes peg-in orders. It loops through each `orderHash` and calls [`_finalizeUnwrap()`](#_finalizeunwrap), which verifies that the order has not already been completed and transfers the token and amount stored in [`unwrapSent`](#unwrapsent) to the recipient. The order is then marked as completed, and the [`FinalizeUnwrapEvent`](#finalizeunwrapevent) is emitted.
```

Updates the contract managing timelocks.

##### Parameters
```solidity
address _timeLock
```
#### `setTimeLockThreshold`

Sets the global timelock threshold.

##### Parameters
```solidity
uint256 _timeLockThreshold
```

#### `setTimeLockThresholdByToken`

Sets a custom timelock threshold per token.

##### Parameters
```solidity
address token, 
uint256 _timeLockThreshold
```

#### `addAllowList`

Adds an address to the allowlist, granting it permission to perform specific contract actions.

#### `removeAllowedList`

Removes an address from the allowlist, revoking its access.

##### Parameters
```solidity
address account
```

#### `pause`
Pauses the contract, preventing token transfers until the contract is unpaused.

#### `unpause`
Resumes contract operations after a pause, allowing bridging and transfers again.

### Read-Only Functions

#### `offAllowList`

Returns `true` if the provided address is not on the allowlist. 

##### Parameters
```solidity
address account
```

#### `onAllowList`

Returns `true` if the provided address is on the allowlist, which means it has permission to use the contract’s functions.

### Relevant Internal Functions

#### `_transfer`

> This internal function is responsible for processing token transfers when a user sends tokens into the bridge. It is called from [`sendMessageWithToken`](#sendmessagewithtoken) and performs validations, calculates and deducts fees, and sends the remaining amount to the receiver. This function ensures the transfer amount is within allowed limits and that it is large enough to cover the minimum fee. If the token is burnable, if burns the amount minus the fee, which is sent to the `BridgeRegistry` contract.

Reviewer: Qué sería "sends the remaining amount to the receiver"?
